test/java/seedu/addressbook/data/AddressBookTest.java

```diff
@@ -0,0 +1,265 @@
+            assertFalse(emptyAddressBook.containsPerson(person));
+        }
+    }
+
+    @Test
+    public void containsTag() throws Exception {
+        UniqueTagList tagsWhichShouldBeIn = new UniqueTagList(tagMathematician, tagScientist);
+        UniqueTagList tagsWHichShouldNotBeIn = new UniqueTagList(tagEconomist, tagPrizeWinner);
+        
+        for (Tag tagWhichShouldBeIn : tagsWhichShouldBeIn) {
+            assertTrue(defaultAddressBook.containsTag(tagWhichShouldBeIn));
+        }
+        for (Tag tagWhichShouldNotBeIn : tagsWHichShouldNotBeIn) {
+            assertFalse(defaultAddressBook.containsTag(tagWhichShouldNotBeIn));
+        }
+        
+        UniqueTagList allTags = new UniqueTagList(tagPrizeWinner, tagScientist, tagMathematician, tagEconomist);
+        
+        for (Tag tag : allTags) {
+            assertFalse(emptyAddressBook.containsTag(tag));
+        }
+    }
+
+    @Test
+    public void removePerson_personExists_removesNormally() throws Exception {
+        int numberOfPersonsBeforeRemoval = getSize(defaultAddressBook.getAllPersons());
+        defaultAddressBook.removePerson(aliceBetsy);
+        
+        assertFalse(defaultAddressBook.containsPerson(aliceBetsy));
+        
+        int numberOfPersonsAfterRemoval = getSize(defaultAddressBook.getAllPersons());
+        assertTrue(numberOfPersonsBeforeRemoval == numberOfPersonsAfterRemoval + 1);
+        
+    }
+
+    @Test
+    public void removePerson_personNotExists_throwsPersonNotFoundException() throws Exception {
+        thrown.expect(PersonNotFoundException.class);
+        defaultAddressBook.removePerson(charlieDouglas);
+    }
+
+    @Test
+    public void removeTag_tagExistsAndUnused_removesNormally() throws Exception {
+        int numberOfTagsBeforeRemoval = getSize(defaultAddressBook.getAllTags());
+        defaultAddressBook.removeTag(tagScientist);
+        
+        assertFalse(defaultAddressBook.containsTag(tagScientist));
+        
+        int numberOfTagsAfterRemoval = getSize(defaultAddressBook.getAllTags());
+        assertTrue(numberOfTagsBeforeRemoval == numberOfTagsAfterRemoval + 1);
+
+    }
+
+    @Test
+    public void removeTag_tagExistsAndUsed_removesUsedTagFromPersons() throws Exception {
+        aliceBetsy.setTags(new UniqueTagList(tagMathematician, tagPrizeWinner));
+        bobChaplin.setTags(new UniqueTagList(tagEconomist));
+        
+        int numberOfTagsBeforeRemoval = getSize(defaultAddressBook.getAllTags());
+        defaultAddressBook.removeTag(tagMathematician);
+        
+        assertTrue(isIdentical(aliceBetsy.getTags(), new UniqueTagList(tagPrizeWinner)));
+        assertTrue(isIdentical(bobChaplin.getTags(), new UniqueTagList(tagEconomist)));
+        assertFalse(defaultAddressBook.containsTag(tagMathematician));
+        
+        int numberOfTagsAfterRemoval = getSize(defaultAddressBook.getAllTags());
+        assertTrue(numberOfTagsBeforeRemoval == numberOfTagsAfterRemoval + 1);
```

Comment:
Is this more intuitive? `numberOfTagsAfterRemoval == numberOfTagsBeforeRemoval - 1`